### PR TITLE
Implement runtime api client side directly in the runtime

### PR DESCRIPTION
**Pull request by bkchr**
This pull requests moves the implementation of the runtime api client side into the runtime as well. This makes the api safer, as the compiler now enforces the user to implement a certain api when he wants to use it. The compile time checks obviously only work for the native runtime interface.
The `Core` api is now also enforced by the compiler by making each other api derive from it.

This pull requests also moves the `sr-api` crate into `substrate-client` and thus, the client is required to compile on `wasm` as well. On `wasm` the client only exposes the api stuff, the rest is disabled. I made this design decision to minimize the size of a where clause when using a runtime api to made the api more ergonomic. Otherwise, the user would needed to specify the error each time and that blowed up easily. I'm still not quite happy with this solution, maybe while reviewing this pull request, we can come to a better solution.

In a follow up pull request, I will provide some sort of macro to auto implement much of the stuff that is now hand written code in the `test-runtime` and `node/runtime`. So, all the `ClientWithApi` stuff should be gone with this follow up pr.

Progress on: https://github.com/paritytech/substrate/issues/1031

### CODE REVIEW

File: core/client/src/block_builder/mod.rs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Utility struct to build a block.
+
+#[cfg(feature = "std")]
+ mod block_builder;
```
Comment (gavofyork) [inline core/client/src/block_builder/mod.rs]: errant space

File: core/client/src/runtime_api/macros.rs [hunk 2]
```diff
@@ -14,33 +14,14 @@
...
-
-#[doc(hidden)]
-pub use primitives::{traits::Block as BlockT, generic::BlockId, transaction_validity::TransactionValidity, ApplyResult};
-use runtime_version::{ApiId, RuntimeVersion};
-use rstd::vec::Vec;
-#[doc(hidden)]
-pub use rstd::slice;
-#[doc(hidden)]
-pub use codec::{Encode, Decode};
+//! Macros for declaring and implementing the runtime API's.
```
Comment (gavofyork) [inline core/client/src/runtime_api/macros.rs]: grocer's apostrophe

**Comment by gavofyork** [comment]
good to hear the `ClientWithApi` stuff will be gone soon.

**Comment by gavofyork** [comment]
would be nice to have an example of a new API being added and used from within `node`. (perhaps not in this PR though)

**Comment by bkchr** [comment]
Fixed the grumbles and the merge conflict. :)
I also created an issue for adding the API example.

File: core/client/src/runtime_api/core.rs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        &self,
+        at: &BlockId<Block>,
+        header: &<Block as BlockT>::Header
+    ) -> Result<()>;
+}
+
+pub mod runtime {
+	use super::*;
+
+    /// The `Core` api trait that is mandantory for each runtime.
```
Comment (gavofyork) [inline core/client/src/runtime_api/core.rs]: spaces!

**Comment by bkchr** [comment]
Everything where `ClientWithApi` is involved in `runtime` and `test-runtime`, will be auto-generated (at least that is the plan). I'm currently looking into it on how that can be done in the best way.

**Comment by bkchr** [comment]
@gavofyork resolved the merge conflict.

**Pull request merged**